exporters/deploytime/app.py

```diff
@@ -67,168 +69,45 @@ def collect(self) -> Iterable[GaugeMetricFamily]:
-            logging.info(
-                "No namespaces specified, watching all namespaces with given PROD_LABEL (%s)",
-                self.prod_label,
-            )
-            query_args = dict(label_selector=self.prod_label)
-        else:
-            logging.info(
-                "No namespaces specified and no PROD_LABEL given, watching all namespaces."
-            )
-            query_args = dict()
-
-        all_namespaces = self.client.resources.get(api_version="v1", kind="Namespace")
-        namespaces = {
-            namespace.metadata.name
-            for namespace in all_namespaces.get(**query_args).items
-        }
-        logging.info("Watching namespaces %s", namespaces)
-        if not namespaces:
-            logging.warning(
-                "No NAMESPACES given and PROD_LABEL did not return any matching namespaces."
-            )
-        return namespaces
-
     def generate_metrics(self) -> Iterable[DeployTimeMetric]:
-        namespaces = self.get_and_log_namespaces()
+        namespaces = get_and_log_namespaces(
+            self.client, self.namespaces, self.prod_label
+        )
+
         if not namespaces:
             return []
 
-        app_label = self.app_label
-        visited_replicas = set()
-
-        def already_seen(full_path: str) -> bool:
-            return full_path in visited_replicas
-
-        def mark_as_seen(full_path: str):
-            visited_replicas.add(full_path)
-
         logging.info("generate_metrics: start")
 
-        # get all running Pods with the app label
-        v1_pods = self.client.resources.get(api_version="v1", kind="Pod")
-        pods = v1_pods.get(
-            label_selector=app_label, field_selector="status.phase=Running"
-        ).items
+        pods = get_running_pods(self.client, namespaces, self.app_label)
 
-        replicas_dict = (
-            self.get_replicas("v1", "ReplicationController")
-            | self.get_replicas("apps/v1", "ReplicaSet")
-            | self.get_replicas("extensions/v1beta1", "ReplicaSet")
-        )
+        # Build dictionary with controllers and retrieved pods
+        replica_pods_dict = filter_pods_by_replica_uid(pods)
 
-        for pod in pods:
-            namespace = pod.metadata.namespace
-            owner_refs = pod.metadata.ownerReferences
-            if namespace not in namespaces or not owner_refs:
-                continue
+        for uid, pod in replica_pods_dict.items():
+            replicas = get_owner_object_from_child(dyn_client, uid, pod)
 
-            logging.debug(
-                "Getting Replicas for pod: %s in namespace: %s",
-                pod.metadata.name,
-                pod.metadata.namespace,
-            )
+            # Since a commit will be built into a particular image and there could be multiple
+            # containers (images) per pod, we will push one metric per image/container in the
+            # pod template
+            images = get_images_from_pod(pod)
```

Comment:
the comment here I think is not clear, can you rewrite it?

also, this function should return  list to be easier to use it

but, shouldn't it check if each image sha is different? (return a set)

Reply:
Return function gives us a `Dict[str, str]` where the first item is a key that corresponds just to the image `sha` and the value is a full URI pointing to a docker image. The `keys()` from the returned dictionary gives us a unique image sha, so there is no need to use set(). Secondly list would be simpler, however we are interested in the full docker URI as well for later usage by other commands such as `skopeo`. I know this is not relevant at this stage, but do want to have function generic enough to give us the flexibility:

```
def get_images_from_pod(pod: ResourceField) -> Dict[str, str]:
    image_shas = {}

......

    for container_status in pod.status.containerStatuses:
        image_shas[
            sha256_value
        ] = f"docker://{registry}{image_name}{sha256_value}"

   return image_shas
```
For example `image_sha : URI`:
```
{
'sha256:6b88f190c7a6cd87c1f1b552c20855e606773cfa96d43f4402840a4b9d2e13a8' :
'docker://quay.io/pelorus/pelorus-committime-exporter@sha256:6b88f190c7a6cd87c1f1b552c20855e606773cfa96d43f4402840a4b9d2e13a8'
}
```

Allows us to have sha256 used in the metric, but as well will allow us to run based on the value URI:

```
skopeo inspect docker://quay.io/pelorus/pelorus-committime-exporter@sha256:6b88f190c7a6cd87c1f1b552c20855e606773cfa96d43f4402840a4b9d2e13a8
```